### PR TITLE
fix: decorate PBS & slurm options with ,omitempty

### PR DESCRIPTION
**Pull request by phillip-gaisford**
## Description
image

Fixing the unit test involved defaulting the batch_args elements to null, which may have implications for dispatcher RM (null pointers). Perhaps there is a better way.

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

**Review by jerryharrow — APPROVED** [review]

**Comment by phillip-gaisford** [comment]
Fixing the unit test involved defaulting the batch_args elements to null, which may have implications for dispatcher RM (null pointers). Perhaps there is a better way.

**Comment by jerryharrow** [comment]
The test case is schemas/test_cases/v0/experiment.yaml

Seems like you should be able to adjust the test case?

- name: minimal valid experiment
  sane_as:
    - http://determined.ai/schemas/expconf/v0/experiment.json
  default_as:
    http://determined.ai/schemas/expconf/v0/experiment.json
  case:
    searcher:
      name: single
      metric: loss
      max_length:
        batches: 1000
    entrypoint: model_def:MyTrial
  #####
  defaulted:
...

    pbs:
      pbsbatch_args: []
...
    slurm:
      sbatch_args: []

**Comment by phillip-gaisford** [comment]
> The test case is schemas/test_cases/v0/experiment.yaml
> 
> Seems like you should be able to adjust the test case?
> 
> * name: minimal valid experiment
>   sane_as:
>   
>   * http://determined.ai/schemas/expconf/v0/experiment.json
>     default_as:
>     http://determined.ai/schemas/expconf/v0/experiment.json
>     case:
>     searcher:
>     name: single
>     metric: loss
>     max_length:
>     batches: 1000
>     entrypoint: model_def:MyTrial
>   
>   
>   defaulted:
>   ...
>   pbs:
>   pbsbatch_args: []
>   ...
>   slurm:
>   sbatch_args: []

I did try that, but found no way of specifying a result that satisfies all test cases.

File: harness/determined/common/schemas/expconf/_gen.py [hunk 1]
```diff
@@ -1480,7 +1480,7 @@
                 "array",
                 "null"
             ],
-            "default": [],
+            "default": null,
```
Comment (jerryharrow) [inline harness/determined/common/schemas/expconf/_gen.py:1483]: Looking at https://pkg.go.dev/encoding/json, this is probably right.   The default value for a struct on omitempty would be a nil, not an empty array.

If the slurm section was present then the sbatch_args default value would be [] though.

> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string.

**Review by jerryharrow — APPROVED** [review]

**Review by rcorujo — APPROVED** [review]
The changes look reasonable, although I don't have intimate knowledge of this space.

**Pull request merged**